### PR TITLE
Speed up mac builds

### LINKED ISSUE #115263: Streamline build time with universal builds
```
FYI: Unfortunately our build just jumped from 59m to 1h33m with the introduction of the universal macOS target :cry:
```

- [ ] Identify the timeline for unifying x86_64, arm64 and universal
- [ ] Have the 3 builds generate from a single job

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
Fixes #115263

This has some minor improvements over https://github.com/microsoft/vscode/pull/116739

that job took 1hr 13min: https://dev.azure.com/monacotools/Monaco/_build/results?buildId=105100&view=results
this change took 54min*: https://dev.azure.com/monacotools/Monaco/_build/results?buildId=105610&view=results

*EDIT: Shoot... I just realized that @deepak1556's change ran the tests and mine didn't... that adds probably another 10min to mine... (so 1hr 4min)

Basically the refactor is to split up signing and build jobs.

Now it works like this:

x64 build
arm build
universal build (waits on x64 and arm)
x64 publish (waits on x64 build)
arm publish (waits on arm build)
universal publish (waits on universal build)

Then we can be sure that the signing is done in parallel. For example, the arm build finishes in 10 min - that can go right to signing when it's done rather than waiting for the universal build to sign everything.

### CODE REVIEW

File: build/azure-pipelines/darwin/product-build-darwin-sign.yml [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    condition: and(succeeded(), ne(variables['VSCODE_ARCH'], 'arm64'))
+
+  - script: |
+      set -e
+      VSCODE_MIXIN_PASSWORD="$(github-distro-mixin-password)" \
+      AZURE_DOCUMENTDB_MASTERKEY="$(builds-docdb-key-readwrite)" \
+      AZURE_STORAGE_ACCESS_KEY="$(ticino-storage-key)" \
+      AZURE_STORAGE_ACCESS_KEY_2="$(vscode-storage-key)" \
+      node build/azure-pipelines/common/createAsset.js \
+        "darwin-$(VSCODE_ARCH)" \
```
Comment (deepak1556) [inline build/azure-pipelines/darwin/product-build-darwin-sign.yml:190]: For x64 it is just `darwin` due to historical reason, changing this key will not push out this update.

x64 - darwin
arm64 - darwin-arm64
universal - darwin-universal

File: build/azure-pipelines/darwin/product-build-darwin-sign.yml [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  - script: |
+      set -e
+      security create-keychain -p pwd $(agent.tempdirectory)/buildagent.keychain
+      security default-keychain -s $(agent.tempdirectory)/buildagent.keychain
+      security unlock-keychain -p pwd $(agent.tempdirectory)/buildagent.keychain
+      echo "$(macos-developer-certificate)" | base64 -D > $(agent.tempdirectory)/cert.p12
+      security import $(agent.tempdirectory)/cert.p12 -k $(agent.tempdirectory)/buildagent.keychain -P "$(macos-developer-certificate-key)" -T /usr/bin/codesign
+      security set-key-partition-list -S apple-tool:,apple:,codesign: -s -k pwd $(agent.tempdirectory)/buildagent.keychain
+      VSCODE_ARCH=$(VSCODE_ARCH) DEBUG=electron-osx-sign* node build/darwin/sign.js
+    displayName: Set Hardened Entitlements
```
Comment (deepak1556) [inline build/azure-pipelines/darwin/product-build-darwin-sign.yml:113]: Hardened entitlements should be set in the build stage otherwise app will crash when running test suite.

Comment (TylerLeonhardt) [reply build/azure-pipelines/darwin/product-build-darwin-sign.yml]: moved

File: build/azure-pipelines/darwin/product-build-darwin-sign.yml (hunk 1 above)
Comment (TylerLeonhardt) [reply build/azure-pipelines/darwin/product-build-darwin-sign.yml]: added case statement

**Comment by TylerLeonhardt** [comment]
So the most recent run of this is still going and it's already at 1hr 30min... which means there's like 30min of variance in this build due to:

* How long it takes for dependencies to be installed
* How long ESRP CodeSign task takes (which we run twice)

I feel like based on my testing that this one, on average, will take less time than @deepak1556's but it won't always be the case... so really I think the PR that should be accepted should be the more readable one.

But maybe @joaomoreno has other opinions.

File: build/azure-pipelines/darwin/product-build-darwin-sign.yml [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    condition: and(succeeded(), ne(variables.NODE_MODULES_RESTORED, 'true'))
+
+  - script: |
+      set -e
+      npx https://aka.ms/enablesecurefeed standAlone
+    timeoutInMinutes: 5
+    condition: and(succeeded(), ne(variables.NODE_MODULES_RESTORED, 'true'), eq(variables['ENABLE_TERRAPIN'], 'true'))
+    displayName: Switch to Terrapin packages
+
+  - script: |
```
Comment (deepak1556) [inline build/azure-pipelines/darwin/product-build-darwin-sign.yml:61]: Now that signing step does not require any fancy node script except for `createAsset`, can you remove the node module cache steps and simplify the yarn step to just compile `src/build` packages.

```
$ cd build
$ yarn && yarn compile
```

File: build/azure-pipelines/darwin/product-build-darwin-sign.yml [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      tar -xzf .build/node_modules_cache/cache.tgz
+    condition: and(succeeded(), eq(variables.NODE_MODULES_RESTORED, 'true'))
+    displayName: Extract node_modules cache
+
+  - script: |
+      set -e
+      npm install -g node-gyp@latest
+      node-gyp --version
+    displayName: Update node-gyp
+    condition: and(succeeded(), ne(variables.NODE_MODULES_RESTORED, 'true'))
```
Comment (deepak1556) [inline build/azure-pipelines/darwin/product-build-darwin-sign.yml:52]: can be removed.

File: build/azure-pipelines/darwin/product-build-darwin-sign.yml [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    displayName: Install dependencies
+    condition: and(succeeded(), ne(variables.NODE_MODULES_RESTORED, 'true'))
+
+  - script: |
+      set -e
+      node build/azure-pipelines/common/listNodeModules.js .build/node_modules_list.txt
+      mkdir -p .build/node_modules_cache
+      tar -czf .build/node_modules_cache/cache.tgz --files-from .build/node_modules_list.txt
+    condition: and(succeeded(), ne(variables.NODE_MODULES_RESTORED, 'true'))
+    displayName: Create node_modules archive
```
Comment (deepak1556) [inline build/azure-pipelines/darwin/product-build-darwin-sign.yml:87]: Can be removed

File: build/azure-pipelines/darwin/product-build-darwin-sign.yml [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        ../VSCode-darwin-$(VSCODE_ARCH).zip
+    displayName: Publish Clients
+
+  - script: |
+      AZURE_STORAGE_ACCESS_KEY="$(ticino-storage-key)" \
+      VSCODE_ARCH="$(VSCODE_ARCH)" \
+      yarn gulp upload-vscode-configuration
+    displayName: Upload configuration (for Bing settings search)
+    condition: and(succeeded(), eq(variables['VSCODE_ARCH'], 'x64'))
+    continueOnError: true
```
Comment (deepak1556) [inline build/azure-pipelines/darwin/product-build-darwin-sign.yml:199]: Bing settings step does not require codesigned build, lets move it to the client build stage.

**Comment by deepak1556** [comment]
Trimming down the yarn step in the signing stage should reduce the timing a bit, lets see how much that improves.

**Comment by TylerLeonhardt** [comment]
Last 2 successful runs has taken 1hr 5min.

**Review by deepak1556 — APPROVED** [review]

**Comment by joaomoreno** [comment]
🥳
